pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java

```diff
@@ -76,6 +80,67 @@ public ImmutableJsonIndexReader(PinotDataBuffer dataBuffer, int numDocs) {
     _docIdMapping = dataBuffer.view(invertedIndexEndOffset, docIdMappingEndOffset, ByteOrder.LITTLE_ENDIAN);
   }
 
+  @Override
+  public String[] getValuesForKeyAndDocs(String key, int[] docIds, Map<Object, RoaringBitmap> cache) {
+    RoaringBitmap docIdMask = RoaringBitmap.bitmapOf(docIds);
+    int[] dictIds = getDictIdsForKey(key);
+    Map<Integer, String> docIdToValues = new HashMap<>(docIds.length);
+
+    if (cache.isEmpty()) {
+      for (int dictId = dictIds[0]; dictId < dictIds[1]; dictId++) {
+        // get docIds from posting list, convert these to the actual docIds
+        ImmutableRoaringBitmap flattenedDocIds = _invertedIndex.getDocIds(dictId);
+        PeekableIntIterator it = flattenedDocIds.getIntIterator();
+        MutableRoaringBitmap realDocIds = new MutableRoaringBitmap();
+        while (it.hasNext()) {
+          realDocIds.add(getDocId(it.next()));
+        }
+        cache.put(dictId, realDocIds.toRoaringBitmap());
+      }
+    }
+
+    for (int dictId = dictIds[0]; dictId < dictIds[1]; dictId++) {
+      RoaringBitmap intersection = RoaringBitmap.and(cache.get(dictId), docIdMask);
+      if (intersection.isEmpty()) {
+        continue;
+      }
+      // dictionary value lookup, stripping the path prefix
+      String val = _dictionary.getStringValue(dictId).substring(key.length() + 1);
+      for (int docId : intersection) {
+        docIdToValues.put(docId, val);
+      }
+    }
+
+    String[] values = new String[docIds.length];
+    for (int i = 0; i < docIds.length; i++) {
+      values[i] = docIdToValues.get(docIds[i]);
```

Comment:
We can iterate over the map instead of the doc ids to fill the values

Reply:
If I iterate over the map I don't know which index of `values[]` corresponds to the `docId` of the current Map.Entry, so I'd need to maintain another mapping from `docId` to the index of the input `docIds[]` array right?